src/backend/renderer/pixman/mod.rs

```diff
@@ -0,0 +1,1183 @@
+            None
+        };
+
+        Ok(TextureAccessor {
+            #[cfg(feature = "wayland_frontend")]
+            buffer,
+            image: &self.0.image,
+            _guard: guard,
+        })
+    }
+}
+
+/// A handle to a pixman texture
+#[derive(Debug, Clone)]
+pub struct PixmanTexture(PixmanImage);
+
+impl From<pixman::Image<'static, 'static>> for PixmanTexture {
+    fn from(image: pixman::Image<'static, 'static>) -> Self {
+        Self(PixmanImage(Rc::new(PixmanImageInner {
+            #[cfg(feature = "wayland_frontend")]
+            buffer: None,
+            dmabuf: None,
+            _flipped: false,
+            image,
+        })))
+    }
+}
+
+struct DmabufReadGuard {
+    dmabuf: Dmabuf,
+}
+
+impl DmabufReadGuard {
+    #[profiling::function]
+    pub fn new(dmabuf: Dmabuf) -> std::io::Result<Self> {
+        dmabuf.sync(DmaBufSyncFlags::START | DmaBufSyncFlags::READ)?;
+        Ok(Self { dmabuf })
+    }
+}
+
+impl Drop for DmabufReadGuard {
+    #[profiling::function]
+    fn drop(&mut self) {
+        if let Err(err) = self.dmabuf.sync(DmaBufSyncFlags::END | DmaBufSyncFlags::READ) {
+            tracing::warn!(?err, "failed to end sync read");
+        }
+    }
+}
+
+struct TextureAccessor<'l> {
+    #[cfg(feature = "wayland_frontend")]
+    buffer: Option<wl_buffer::WlBuffer>,
+    image: &'l Image<'static, 'static>,
+    _guard: Option<DmabufReadGuard>,
+}
+
+impl<'l> TextureAccessor<'l> {
+    fn with_image<F, R>(&self, f: F) -> Result<R, PixmanError>
+    where
+        F: FnOnce(&'l Image<'static, 'static>) -> R,
+    {
+        #[cfg(feature = "wayland_frontend")]
+        if let Some(buffer) = self.buffer.as_ref() {
+            // We only have a buffer in case the image was created from
+            // a shm buffer. In this case we need to guard against SIGPIPE
+            // when accessing the image
+            return Ok(shm::with_buffer_contents(buffer, |_, _, _| f(self.image))?);
+        }
+
+        Ok(f(self.image))
+    }
+}
+
+impl PixmanTexture {
+    #[profiling::function]
+    fn accessor<'l>(&'l self) -> Result<TextureAccessor<'l>, PixmanError> {
+        self.0.accessor()
+    }
+}
+
+impl Texture for PixmanTexture {
+    fn width(&self) -> u32 {
+        self.0 .0.image.width() as u32
+    }
+
+    fn height(&self) -> u32 {
+        self.0 .0.image.height() as u32
+    }
+
+    fn format(&self) -> Option<DrmFourcc> {
+        DrmFourcc::try_from(self.0 .0.image.format()).ok()
+    }
+}
+
+/// Handle to the currently rendered frame during [`PixmanRenderer::render`](Renderer::render).
+#[derive(Debug)]
+pub struct PixmanFrame<'frame> {
+    renderer: &'frame mut PixmanRenderer,
+
+    transform: Transform,
+    output_size: Size<i32, Physical>,
+    size: Size<i32, Physical>,
+
+    finished: AtomicBool,
+}
+
+impl<'frame> PixmanFrame<'frame> {
+    fn draw_solid_color(
+        &mut self,
+        dst: Rectangle<i32, Physical>,
+        damage: &[Rectangle<i32, Physical>],
```

Comment:
For Smithay's case, does forcing the user to specify damage make sense?

Reply:
Yeah, this is the contract for damage tracking. The renderer is not allowed to draw outside of the supplied damage rect(s) or otherwise you will get artifacts. This can also occur for a partial occluded solid element when the element above is transparent and triggers a redraw.